### PR TITLE
Restore early SyntaxError for strict mode assignment to eval/arguments

### LINKED ISSUE #1267: Rationale behind early ReferenceError and early SyntaxError
I'm attempting to update JSC's parser to accept early `ReferenceError` concept.
But I don't get the point of the rationale behind early ReferenceError and SyntaxError yet.

According to the spec[1,2],

```js
function test() {
    "use strict";
    arguments = 42;
}
```

should raise early ReferenceError. On the other hand,

```js
function test() {
    "use strict";
    ({ arguments } = object);
}
```

```js
function test() {
    "use strict";
    ({ a: arguments = 20 } = object);
}
```

should raise early SyntaxError[3].

What makes this differences? Why are IdentifierReference in Object/ArrayPatterns handled differently from usual AssignmentExpression LHS's IdentifierReference?
This is a bit confusing and it seems that even test262 does not test it correctly (see `test/language/expressions/assignment/id-arguments-strict.js` for example).

1: https://tc39.github.io/ecma262/#sec-assignment-operators-static-semantics-early-errors
2: https://tc39.github.io/ecma262/#sec-identifiers-static-semantics-isvalidsimpleassignmenttarget
3: https://tc39.github.io/ecma262/#sec-destructuring-assignment-static-semantics-early-errors

### PR DESCRIPTION
**Pull request by anba**
Fixes https://bugs.ecmascript.org/show_bug.cgi?id=4375

Closes #1267.

### CODE REVIEW

**Comment by anba** [comment]
https://github.com/IgnoredAmbience/archives/blob/d13266f6be27daa61605bdcd9fe1e6b7ae3ce460/bugzilla/4375.xml

File: spec.html [hunk 1]
```diff
@@ -14451,13 +14457,19 @@ <h1>Static Semantics: Early Errors</h1>
           It is a Syntax Error if |LeftHandSideExpression| is either an |ObjectLiteral| or an |ArrayLiteral| and |LeftHandSideExpression| is not covering an |AssignmentPattern|.
         </li>
         <li>
-          It is an early Reference Error if |LeftHandSideExpression| is neither an |ObjectLiteral| nor an |ArrayLiteral| and IsValidSimpleAssignmentTarget of |LeftHandSideExpression| is *false*.
+          It is an early Reference Error if |LeftHandSideExpression| is neither an |ObjectLiteral| nor an |ArrayLiteral| and IsValidSimpleAssignmentTarget of |LeftHandSideExpression| is ~invalid~.
```
Comment (ljharb) [inline spec.html]: It seems strange to have something named as a predicate (IsValidSimpleAssignmentTarget) not be a boolean. Should this be renamed to something else?

Comment (leobalter) [reply spec.html]: `SimpleAssignmentTarget`, `AssignmentTarget{Type,Hint}` should do the trick or whatever @anba decides works for me as well.

For anything else, this PR LGTM. +1

**Comment by littledan** [comment]
Thanks for following up on this old issue and helping us get aligned with web reality. LGTM, seems like an accurate fix of the linked bug. Test262 tests would be great.

**Comment by anba** [comment]
> Test262 tests would be great.

Already covered by:
<details>
<ul>
<li>test/language/arguments-object/10.5-1-s.js
<li>test/language/arguments-object/10.5-1gs.js
<li>test/language/arguments-object/10.5-7-b-1-s.js
<li>test/language/directive-prologue/14.1-4gs.js
<li>test/language/directive-prologue/14.1-5gs.js
<li>test/language/expressions/assignment/dstr-array-elem-target-simple-strict.js
<li>test/language/expressions/assignment/dstr-obj-id-init-simple-strict.js
<li>test/language/expressions/assignment/dstr-obj-id-simple-strict.js
<li>test/language/expressions/assignment/id-arguments-strict.js
<li>test/language/expressions/assignment/id-eval-strict.js
<li>test/language/expressions/compound-assignment/11.13.2-6-1-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-10-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-11-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-12-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-13-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-14-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-15-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-16-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-17-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-18-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-19-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-1gs.js
<li>test/language/expressions/compound-assignment/11.13.2-6-2-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-20-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-21-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-22-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-3-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-4-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-5-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-6-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-7-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-8-s.js
<li>test/language/expressions/compound-assignment/11.13.2-6-9-s.js
<li>test/language/expressions/postfix-decrement/11.3.2-2-1-s.js
<li>test/language/expressions/postfix-decrement/11.3.2-2-2-s.js
<li>test/language/expressions/postfix-increment/11.3.1-2-1-s.js
<li>test/language/expressions/postfix-increment/11.3.1-2-1gs.js
<li>test/language/expressions/postfix-increment/11.3.1-2-2-s.js
<li>test/language/expressions/prefix-decrement/11.4.5-2-1-s.js
<li>test/language/expressions/prefix-decrement/11.4.5-2-2-s.js
<li>test/language/expressions/prefix-decrement/11.4.5-2-2gs.js
<li>test/language/expressions/prefix-increment/11.4.4-2-1-s.js
<li>test/language/expressions/prefix-increment/11.4.4-2-2-s.js
<li>test/language/statements/for-await-of/async-func-decl-dstr-array-elem-target-simple-strict.js
<li>test/language/statements/for-in/dstr-array-elem-target-simple-strict.js
<li>test/language/statements/for-in/dstr-obj-id-init-simple-strict.js
<li>test/language/statements/for-in/dstr-obj-id-simple-strict.js
<li>test/language/statements/for-of/dstr-array-elem-target-simple-strict.js
<li>test/language/statements/for-of/dstr-obj-id-init-simple-strict.js
<li>test/language/statements/for-of/dstr-obj-id-simple-strict.js
<li>test/language/statements/function/13.0-13-s.js
<li>test/language/statements/function/13.0-14-s.js
<li>test/language/statements/function/13.0-15-s.js
<li>test/language/statements/function/13.0-16-s.js
<li>test/language/statements/function/13.0-7-s.js
<li>test/language/statements/function/13.0-8-s.js
<li>test/language/statements/function/13.0_4-5gs.js
<li>test/language/statements/function/enable-strict-via-body.js
<li>test/language/statements/function/enable-strict-via-outer-body.js
<li>test/language/statements/function/enable-strict-via-outer-script.js
<li>test/language/statements/variable/12.2.1-13-s.js
<li>test/language/statements/variable/12.2.1-15-s.js
<li>test/language/statements/variable/12.2.1-19-s.js
<li>test/language/statements/variable/12.2.1-2-s.js
<li>test/language/statements/variable/12.2.1-4-s.js
<li>test/language/statements/variable/12.2.1-8-s.js
</ul>
</details>

**Comment by littledan** [comment]
OK, it seems like this was an accidental change vs previous versions, corresponds to implementation consensus and tests, and simply restores things from before the typo. I think we should consider this patch ready to land.

File: spec.html [hunk 2]
```diff
@@ -13064,7 +13064,10 @@ <h1>Static Semantics: Early Errors</h1>
       </emu-grammar>
       <ul>
         <li>
-          It is an early Reference Error if IsValidSimpleAssignmentTarget of |LeftHandSideExpression| is *false*.
+          It is an early Reference Error if IsValidSimpleAssignmentTarget of |LeftHandSideExpression| is ~invalid~.
```
Comment (ljharb) [inline spec.html]: This still needs to be renamed to not have an “is” prefix` on an enum

File: spec.html (hunk 1 above)
Comment (bterlson) [reply spec.html]: AssignTargetType: +1

File: spec.html (hunk 2 above)
Comment (ljharb) [reply spec.html]: in the editor's call, we decided to rename this to "AssignmentTargetType". I'll take care of this as part of the rebase and merge.

File: spec.html [hunk 3]
```diff
@@ -12684,7 +12687,10 @@ <h1>Static Semantics: Early Errors</h1>
       </emu-grammar>
       <ul>
         <li>
-          It is an early Reference Error if IsValidSimpleAssignmentTarget of |UnaryExpression| is *false*.
+          It is an early Reference Error if AssignmentTargetType of |UnaryExpression| is ~invalid~.
+        </li>
+        <li>
+          It is an early Syntax Error if AssignmentTargetType of |LeftHandSideExpression| is ~strict~.
```
Comment (adrianheine) [inline spec.html]: See #1423.

**Pull request merged**